File: test_samples/comprehensive/Complex.cs
// Comprehensive C# test file with complex structures
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;

namespace Authentication.Complex
{
    /// <summary>
    /// Authenticator interface
    /// </summary>
    public interface IAuthenticator
    {
        bool Authenticate(string username, string password);
        void Logout();
    }
    
    /// <summary>
    /// Base User class
    /// </summary>
    public abstract class User
    {
        protected string Username { get; set; }
        protected int Id { get; set; }
        
        public User(string username, int id)
        {
            Username = username;
            Id = id;
        }
        
        public abstract string GetRole();
    }
    
    /// <summary>
    /// AdminUser with interface implementation
    /// </summary>
    public class AdminUser : User, IAuthenticator
    {
        private List<string> _permissions;
        private readonly ReaderWriterLockSlim _lock;
        
        public AdminUser(string username, int id) : base(username, id)
        {
            _permissions = new List<string>();
            _lock = new ReaderWriterLockSlim();
        }
        
        public override string GetRole()
        {
            return "ADMIN";
        }
        
        public bool Authenticate(string username, string password)
        {
            if (string.IsNullOrEmpty(username) || string.IsNullOrEmpty(password))
            {
                return false;
            }
            return Username == username && ValidatePassword(password);
        }
        
        public void Logout()
        {
            Console.WriteLine("Admin logged out");
        }
        
        public void AddPermission(string permission)
        {
            _lock.EnterWriteLock();
            try
            {
                _permissions.Add(permission);
            }
            finally
            {
                _lock.ExitWriteLock();
            }
        }
        
        private bool ValidatePassword(string password)
        {
            return password.Length >= 8;
        }
    }
    
    /// <summary>
    /// Generic SessionManager
    /// </summary>
    public class SessionManager<T> where T : User
    {
        private static SessionManager<T> _instance;
        private static readonly object _lock = new object();
        private Dictionary<string, T> _sessions;
        
        private SessionManager()
        {
            _sessions = new Dictionary<string, T>();
        }
        
        public static SessionManager<T> GetInstance()
        {
            if (_instance == null)
            {
                lock (_lock)
                {
                    if (_instance == null)
                    {
                        _instance = new SessionManager<T>();
                    }
                }
            }
            return _instance;
        }
        
        public void CreateSession(string sessionId, T user)
        {
            lock (_lock)
            {
                _sessions[sessionId] = user;
            }
        }
        
        public T GetSession(string sessionId)
        {
            lock (_lock)
            {
                return _sessions.TryGetValue(sessionId, out var user) ? user : null;
            }
        }
    }
    
    /// <summary>
    /// StatusCode enum
    /// </summary>
    public enum StatusCode
    {
        Success = 200,
        Unauthorized = 401,
        Forbidden = 403,
        NotFound = 404
    }
    
    /// <summary>
    /// Extension methods for StatusCode
    /// </summary>
    public static class StatusCodeExtensions
    {
        public static string GetMessage(this StatusCode code)
        {
            return code switch
            {
                StatusCode.Success => "OK",
                StatusCode.Unauthorized => "Unauthorized",
                StatusCode.Forbidden => "Forbidden",
                StatusCode.NotFound => "Not Found",
                _ => "Unknown"
            };
        }
    }
    
    /// <summary>
    /// Complex authentication service
    /// </summary>
    public class AuthenticationService
    {
        public (bool success, string sessionId) AuthenticateAndCreateSession<T>(
            T user,
            string username,
            string password,
            SessionManager<T> manager
        ) where T : User, IAuthenticator
        {
            if (user.Authenticate(username, password))
            {
                var sessionId = Guid.NewGuid().ToString();
                manager.CreateSession(sessionId, user);
                return (true, sessionId);
            }
            return (false, null);
        }
    }
}

File: test_samples/comprehensive/complex.rb
# Comprehensive Ruby test file with complex structures

module Authentication
  # Authenticator module
  module Authenticator
    def authenticate(username, password)
      raise NotImplementedError
    end
    
    def logout
      raise NotImplementedError
    end
  end
  
  # Base User class
  class User
    attr_reader :username, :id
    
    def initialize(username, id)
      @username = username
      @id = id
      @mutex = Mutex.new
    end
    
    def role
      raise NotImplementedError
    end
  end
  
  # AdminUser with mixed-in authenticator
  class AdminUser < User
    include Authenticator
    
    attr_accessor :permissions
    
    def initialize(username, id)
      super(username, id)
      @permissions = []
    end
    
    def role
      'ADMIN'
    end
    
    def authenticate(username, password)
      return false if username.nil? || password.nil?
      @username == username && valid_password?(password)
    end
    
    def logout
      puts 'Admin logged out'
    end
    
    def add_permission(permission)
      @mutex.synchronize do
        @permissions << permission
      end
    end
    
    private
    
    def valid_password?(password)
      password.length >= 8
    end
  end
  
  # Generic SessionManager class
  class SessionManager
    def initialize
      @sessions = {}
      @mutex = Mutex.new
    end
    
    def create_session(session_id, user)
      @mutex.synchronize do
        @sessions[session_id] = user
      end
    end
    
    def get_session(session_id)
      @mutex.synchronize do
        @sessions[session_id]
      end
    end
    
    def remove_session(session_id)
      @mutex.synchronize do
        @sessions.delete(session_id)
      end
    end
    
    # Singleton pattern
    class << self
      attr_accessor :instance
      
      def get_instance
        @instance ||= new
      end
    end
  end
  
  # Status code module
  module StatusCode
    SUCCESS = { code: 200, message: 'OK' }.freeze
    UNAUTHORIZED = { code: 401, message: 'Unauthorized' }.freeze
    FORBIDDEN = { code: 403, message: 'Forbidden' }.freeze
    NOT_FOUND = { code: 404, message: 'Not Found' }.freeze
    
    def self.get_message(code)
      constants.each do |const_name|
        const = const_get(const_name)
        return const[:message] if const[:code] == code
      end
      'Unknown'
    end
  end
end
